### PR TITLE
set 500ms debounce timer on graph refresh

### PR DESCRIPTION
**Pull request by stuartnelson3**
requests to refresh are now queued in a timer. if an additional request is received before the timer expires, the original timer is canceled and a new one is created.

canceling an expired timer has no negative effects.

the 500ms debounce is arbitrary. Let me know what you guys think it should be set to. I originally tried 200ms but it was too quick, and would often reload between clicks.

@juliusv @grobie 

requested in #287

### CODE REVIEW

**Comment by grobie** [comment]
This should fix #153. @matthiasr

File: app/assets/javascripts/angular/controllers/graph_ctrl.js [hunk 1]
```diff
@@ -123,10 +126,15 @@ angular.module("Prometheus.controllers").controller('GraphCtrl',
       var rangeSeconds = Prometheus.Graph.parseDuration(scope.graph.range);
       // bigger denominator == smaller step == more data
       var step = Math.floor(rangeSeconds / scalingFactor)
-      refreshFn(rangeSeconds, step).then(function(data) {
+
+      // Cancels the reload request if it exists.
+      $timeout.cancel(debounce);
+      debounce = $timeout(function() {
+        refreshFn(rangeSeconds, step).then(function(data) {
         scope.$broadcast('redrawGraphs', data);
```
Comment (juliusv) [inline app/assets/javascripts/angular/controllers/graph_ctrl.js]: missing indent for these lines?

File: app/assets/javascripts/angular/controllers/graph_ctrl.js [hunk 2]
```diff
@@ -123,10 +126,15 @@ angular.module("Prometheus.controllers").controller('GraphCtrl',
...
         AnnotationRefresher(scope.graph, scope);
-      });
+        });
+      }, 500);
```
Comment (juliusv) [inline app/assets/javascripts/angular/controllers/graph_ctrl.js]: What concerns me is that this 500ms debounce is now added to any and all graph loads, not just the ones that are triggered by time adjustments. This includes initial page load, as well as any other manual changes to graphs which don't profit from being debounced. One solution I can think of is to pass a boolean `debounce` argument to `refreshGraph()`, which would only be set to `true` when the refresh originates from a time settings change, namely in these lines:

https://github.com/prometheus/promdash/blob/debounce/app/assets/javascripts/angular/controllers/graph_ctrl.js#L94-L102

**Comment by stuartnelson3** [comment]
roger, debounce updated to only delay on changes to global time/date settings

**Pull request merged**